chalk-ir/src/fold/boring_impls.rs

```diff
@@ -12,22 +12,23 @@ use std::sync::Arc;
 impl<'a, T: Fold<I, TI>, I: Interner, TI: TargetInterner<I>> Fold<I, TI> for &'a T {
     type Result = T::Result;
     fn fold_with<'i>(
-        &self,
+        self,
         folder: &mut dyn Folder<'i, I, TI>,
         outer_binder: DebruijnIndex,
     ) -> Fallible<Self::Result>
     where
         I: 'i,
         TI: 'i,
     {
-        (**self).fold_with(folder, outer_binder)
+        unimplemented!();
```

Comment:
Heads up: I think implementation should be removed, but I haven't quite figured how to do that.

Reply:
So this is a bit of a weird case, since `self.fold_with()` would call this impl recursively.
This should be `self.clone.fold_with()` instead.

Reply:
I think `(*self).clone()`, but it fails to build because of some missing `Clone` bounds.

Reply:
```
   1: chalk_ir::fold::boring_impls::<impl chalk_ir::fold::Fold<I,TI> for &T>::fold_with
             at ./chalk-ir/src/fold/boring_impls.rs:23:9
   2: <T as chalk_ir::fold::shift::Shift<I>>::shifted_in_from
             at ./chalk-ir/src/fold/shift.rs:32:9
   3: <T as chalk_ir::fold::shift::Shift<I>>::shifted_in
             at ./chalk-ir/src/fold/shift.rs:28:9
   4: chalk_solve::coherence::solve::<impl chalk_solve::coherence::CoherenceSolver<I>>::disjoint::{{closure}}
             at ./chalk-solve/src/coherence/solve.rs:104:26
```